### ISSUE TITLE
E308 mismatched types false positives for infix operators on differing operand types

### ISSUE THREAD

**Issue by rkarp** [intellij-rust/intellij-rust#4081]
## Environment

* **Intellij-Rust plugin version:** 0.2.101.2670-191-nightly (same for latest stable)
* **Rust toolchain version:** 1.35.0
* **IDE name and version:** IDEA 2019.1.3
* **Operating system:** Windows 7 64bit

## Problem description
The type checker doesn't seem to use all trait impls for infix operators when typechecking. Simple cases where both operands are the same type are typechecked correctly, though.

```rust
fn main() {
    let x: i32 = 1;
    let y: i32 = 2;
    // no error
    println!("{}", x + y);

    let x: i32 = 1;
    let y: &i32 = &2;
    // mismatched types [E0308]expected `i32`, found `&i32`
    println!("{}", x + y);

    let x: u64 = 1;
    let y: usize = 2;
    // mismatched types [E0308]expected `u64`, found `usize`
    println!("{}", x << y);
}
```

This seems to be a regression, the errors weren't there a few stable releases ago.

**Comment by lovasoa** [comment]
I was going to open a new issue when I found this one. I've been bitten by that too !

![image](https://user-images.githubusercontent.com/552629/69482433-9c746100-0e1b-11ea-835e-bcb1802ddee2.png)

intellij-rust even (incorrectly) infers the type of integer literals to match the type of the other operand : 

![image](https://user-images.githubusercontent.com/552629/69482442-b615a880-0e1b-11ea-811a-87f472e2a167.png)

**Comment by josephglanville** [comment]
Also got bitten by this. Pretty annoying bug when one is writing a lot of bit twiddling code.

**Comment by Maksim20023** [comment]
I'm closing this ticket due to the fact that the problem does not reproduce in the current environment (IntelliJ IDEA 2023.1.4, Rust plugin 0.4.198.5409-231). If this issue still reproduces for you, please reopen the ticket.
